File: Python/examples/fixed_channel_inprocess.py
from __future__ import annotations

import argparse
import os
import struct
import time

from _common import cleanup_shm, load_xproc, wait_for_message


MESSAGE_COUNT = 10


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--module-dir")
    args = parser.parse_args()

    xproc = load_xproc(args.module_dir)

    shm_path = f"/xproc_python_fixed_inprocess_{os.getpid()}"
    cleanup_shm(xproc, shm_path)

    create_options = xproc.TransportOptions()
    create_options.path = shm_path
    create_options.shm_size = xproc.shm_size_for_data_capacity(16_384)
    create_options.channel_type = xproc.ChannelType.FIXED
    create_options.item_size = 4
    create_options.create_if_missing = True

    attach_options = xproc.TransportOptions()
    attach_options.path = shm_path
    attach_options.shm_size = xproc.INFER_EXISTING_SHM_SIZE
    attach_options.channel_type = xproc.ChannelType.FIXED
    attach_options.item_size = 4
    attach_options.create_if_missing = False

    producer = xproc.Producer(create_options)
    consumer = xproc.Consumer(attach_options)

    expected = 1

    try:
        for value in range(1, MESSAGE_COUNT + 1):
            producer.send_fixed_sized(struct.pack("<i", value))
            time.sleep(0.01)

            payload = wait_for_message(consumer.poll_copy, label="fixed-size payload")
            received = struct.unpack("<i", payload)[0]
            print(f"recv: {received}")
            if received != expected:
                raise RuntimeError(f"sequence mismatch, expected {expected} got {received}")
            expected += 1
    finally:
        consumer.close()
        producer.close()
        cleanup_shm(xproc, shm_path)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: node/examples/fixed_channel_inprocess.ts
const timers = require("node:timers/promises");

const xproc = require("../index.js") as XprocModule;

const delay = timers.setTimeout as (ms: number) => Promise<void>;
const MESSAGE_COUNT = 10;

function cleanupShm(path: string): void {
  try {
    xproc.shmUnlink(path);
  } catch {
    // Best-effort cleanup only.
  }
}

async function main(): Promise<void> {
  const shmPath = `/xproc_node_fixed_inprocess_${process.pid}`;
  cleanupShm(shmPath);

  const createOptions: TransportOptions = {
    path: shmPath,
    shmSize: xproc.shmSizeForDataCapacity(16384n),
    channelType: "fixed",
    itemSize: 4,
    createIfMissing: true,
  };

  const attachOptions: TransportOptions = {
    ...createOptions,
    shmSize: xproc.XPROC_C_INFER_EXISTING_SHM_SIZE,
    createIfMissing: false,
  };

  const producer = new xproc.Producer(createOptions);
  const consumer = new xproc.Consumer(attachOptions);

  let expected = 1;

  try {
    for (let i = 1; i <= MESSAGE_COUNT; ++i) {
      const payload = Buffer.alloc(4);
      payload.writeInt32LE(i, 0);
      producer.sendFixedSized(payload);
      await delay(10);

      while (true) {
        const message = consumer.pollCopy();
        if (message === null) {
          await delay(1);
          continue;
        }

        const value = Buffer.from(message).readInt32LE(0);
        console.log(`recv: ${value}`);
        if (value !== expected) {
          throw new Error(`sequence mismatch, expected ${expected} got ${value}`);
        }
        expected += 1;
        break;
      }
    }
  } finally {
    consumer.close();
    producer.close();
    cleanupShm(shmPath);
  }
}

void main().catch((error: unknown) => {
  const message = error instanceof Error ? error.stack ?? error.message : String(error);
  console.error(message);
  process.exit(1);
});
